cns/kubecontroller/pod/reconciler.go

```diff
@@ -2,79 +2,99 @@ package pod
 
 import (
 	"context"
+	"strconv"
 
 	"github.com/pkg/errors"
+	"go.uber.org/zap"
 	v1 "k8s.io/api/core/v1"
-	"k8s.io/apimachinery/pkg/fields"
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/event"
 	"sigs.k8s.io/controller-runtime/pkg/predicate"
 	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 )
 
-type podcli interface {
+type cli interface {
 	List(ctx context.Context, list client.ObjectList, opts ...client.ListOption) error
 }
 
-type podListener interface {
-	Update([]v1.Pod)
+// watcher watches Pods on the current Node and notifies listeners of changes.
+type watcher struct {
+	z              *zap.Logger
+	cli            cli
+	reconcileFuncs []reconcile.Func
 }
 
-type PodWatcher struct {
-	cli            podcli
-	listOpt        client.ListOption
-	ReconcileFuncs []reconcile.Func
+func New(z *zap.Logger) *watcher { //nolint:revive // force usage of new by keeping the struct private
+	return &watcher{
+		z: z.With(zap.String("component", "pod-watcher")),
+	}
 }
 
-func New(nodename string) *PodWatcher { //nolint:revive // private struct to force constructor
-	return &PodWatcher{
-		listOpt: &client.ListOptions{FieldSelector: fields.SelectorFromSet(fields.Set{"spec.nodeName": nodename})},
-	}
+// With adds reconcile.Funcs to the Watcher.
+func (p *watcher) With(fs ...reconcile.Func) *watcher {
+	p.reconcileFuncs = append(p.reconcileFuncs, fs...)
+	return p
 }
 
-func (p *PodWatcher) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	for _, f := range p.ReconcileFuncs {
-		if _, err := f(ctx, req); err != nil {
-			return reconcile.Result{}, errors.Wrap(err, "failed to reconcile")
+func (p *watcher) Reconcile(ctx context.Context, req reconcile.Request) (ctrl.Result, error) {
+	for _, f := range p.reconcileFuncs {
+		if res, err := f(ctx, req); !res.IsZero() || err != nil {
+			return res, errors.Wrap(err, "failed to reconcile")
```

Comment:
I personally haven't found a good use for a non-empty `reconcile.Result`, but in my mind the intended usage is to indicate "everything was fine, but we want to reconcile this again later". I wonder if it makes sense here to return on the first non-empty `reconcile.Result` (current implementation) or allow all the funcs to run as long as there is no error, saving all the `reconcile.Result`s along the way and returning a non-zero one should one exist. Then you'd kind of have to decide which non-empty one to return which isn't ideal, but if they are indented to be used how I think, I think it doesn't make sense to return early unless `err` is `nil`.

Reply:
hm, I have also never used a non-zero reconcile.Result. I wrote this thinking that you would use it _with_ an error to indicate to not retry immediately, but at a certain point in time in the future. Your interpretation seems more correct...